### DISCUSSION TITLE
Using HTTP listnener to serve a web application

### DISCUSSION THREAD

**Discussion by DUROUXG** [nextgenhealthcare/connect#5944] [Q&A]
Hi I want to use mirth to serve an app svelte made. 1 index.html and couple of .js / .css /.png to send . It works well when i use Static resources by the book. it is to say use HTTP channel Static resources and setup  Context Path: /static/myfile ,  Resource Type: File , Value: /path/to/your/file.html , Content Type: text/plain.  This works but need to have all file hard coded,   I want to stay dynamic by introducing variable  such as ${file}  in my path to catch requested files and send them . it looks to fail. Then I try to find a way to generate a custom HTTP response or catch wrong response to add file in unsuccesfully . I managed partially but the PNG picture was not sent as expected unless declared in Static resources. 

So far i didn't find the right way to customise my response . Is there any method to generate a proper HTTP full response including all file ? 

Thanks

**Comment by DUROUXG** [comment]
Let me clarify my thoughts , here is the code i use : 

var filecontext = $('ReqContPath')
var filePath = 'C://codeGD//svelte//GUItestv1//dist' + filecontext ; 

var fileContent;
var filecontentbin;
try {
   // read in binary for picture or txt for css html js 
    fileContent = FileUtil.read(filePath);
    filecontentbin = FileUtil.readBytes(filePath);
} catch (e) {
 
var r = ResponseFactory.getErrorResponse(''); // empty payload
responseMap.put('custResponse', r);
return;
}
...
// once mime define place it ivariable to be placed in Content-Type
channelMap.put('mime',mime)
}
if (bin){
	var r = ResponseFactory.getSentResponse(FileUtil.encode(filecontentbin));
}else {
var r = ResponseFactory.getSentResponse(fileContent);
}
responseMap.put('custResponse', r);
`
This works but not for images and fonts . However if i use Static Resources instead it works. So i do not understand how i can reproduce the Static resource bahaviour within a js coded similar thing ? is it just possible ? 


here is my chanel setup. 
![image](https://github.com/nextgenhealthcare/connect/assets/41985869/6dc78451-264c-49f9-8a09-722d9516ce11)

**Comment by jonbartels** [comment]
I've done this before for VERY basic webapps in Mirth Connect. 

Mirth Connect *can* do this but it really is not the best tool for the job. 

Some key points:

1. Use **XHTML** and not HTML - XHTML is XML, Mirth does very well at building and returning XML
2. For images you can use a [data URL](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs) - This lets you look up or generate information, images, etc. in Mirth Connect and just render the base64 content. The drawback is that this probably won't cache well and your response XHTML will be massive since it embedsd the images
3. If you have more than one step for your apps, learn how to use `$(d##)` references to read responses from each destination. You can use this to create filters to correctly skip destinations. Then you can have each destination handle one step or request for your application. 
4. Building on item 3 - Each separate destination should then build the response to THE SAME responseMap value. Your source connector will ALWAYS respond from this same responseMap entry regardless of what response ran
5. Every request should have something visible to the user in the XHTML that contains the Mirth Message ID and the timestamp. That way if the user has a problem, they can tell you that information for triage and diagnosis.

**Reply by DUROUXG** [reply]
thanks I will investigate your answer and try it . I will let you know the outcome.
